integrations/tensorflow/iree_tf_compiler/TFL/LowerGlobalTensors.cpp

```diff
@@ -0,0 +1,162 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "iree/compiler/Utils/ConversionUtils.h"
+#include "iree_tf_compiler/TFL/Passes.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "tensorflow/compiler/mlir/lite/ir/tfl_ops.h"
+
+namespace mlir {
+namespace iree_integrations {
+namespace TFL {
+
+class LowerGlobalTensorsPass
+    : public PassWrapper<LowerGlobalTensorsPass, OperationPass<ModuleOp>> {
+ public:
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<mlir::TFL::TensorFlowLiteDialect,
+                    iree_compiler::IREE::Util::UtilDialect>();
+  }
+
+  StringRef getArgument() const override {
+    return "iree-tflite-lower-global-tensors";
+  }
+
+  StringRef getDescription() const override {
+    return "Lowers tflite global tensors to IREE flow dialect variables";
+  }
+
+  // Validates that no TFLite frontends ops are in the module.
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    mlir::OpBuilder builder(moduleOp.body());
+
+    DenseMap<StringRef, FuncOp> symNameToFunction;
+    for (auto func : moduleOp.getOps<FuncOp>()) {
+      symNameToFunction[func.sym_name()] = func;
+    }
+
+    DenseMap<StringRef, DenseElementsAttr> sharedNameToConstant;
+    DenseMap<StringRef, LocationAttr> sharedNameToLoc;
+
+    llvm::SmallVector<mlir::TFL::VarHandleOp> handleOps;
+    llvm::SmallVector<mlir::TFL::AssignVariableOp> assignOps;
+    llvm::SmallVector<mlir::TFL::ReadVariableOp> readOps;
```

Comment:
Any opinions on the expected number of these ops?

https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

> In the absence of a well-motivated choice for the number of inlined elements N, it is recommended to use SmallVector<T> (that is, omitting the N). This will choose a default number of inlined elements reasonable for allocation on the stack (for example, trying to keep sizeof(SmallVector<T>) around 64 bytes).

Reply:
I added some high level examples but this is directly related to the number of variables in the model. In most cases where there are any there will likely be multiple (e.g. LSTMs have multiple state variables and often multiple LSTMs). So I double any fixed value will be helpful.